src/hyperlight_host/src/hypervisor/gdb/hyp_debug.rs

```diff
@@ -0,0 +1,481 @@
+/// Bit position of HW breakpoints status in DR6 debug register
+const DR6_HW_BP_FLAGS_POS: usize = 0;
+/// Bit mask of HW breakpoints status in DR6 debug register
+const DR6_HW_BP_FLAGS_MASK: u64 = 0x0F << DR6_HW_BP_FLAGS_POS;
+
+/// This trait is used to define a common way of interacting with a vCPU to
+/// allow debugging functionality
+pub trait GuestVcpuDebug {
+    /// Type that wraps the vCPU functionality
+    type Vcpu;
+
+    /// Returns true whether the provided address is a hardware breakpoint
+    fn is_hw_breakpoint(&self, addr: &u64) -> bool;
+    /// Stores the address of the hw breakpoint
+    fn save_hw_breakpoint(&mut self, addr: &u64) -> bool;
+    /// Deletes the address of the hw breakpoint from storage
+    fn delete_hw_breakpoint(&mut self, addr: &u64);
+
+    /// Read registers
+    fn read_regs(&self, vcpu_fd: &Self::Vcpu, regs: &mut X86_64Regs) -> Result<()>;
+    /// Enables or disables stepping and sets the vCPU debug configuration
+    fn set_single_step(&mut self, vcpu_fd: &Self::Vcpu, enable: bool) -> Result<()>;
+    /// Translates the guest address to physical address
+    fn translate_gva(&self, vcpu_fd: &Self::Vcpu, gva: u64) -> Result<u64>;
+    /// Write registers
+    fn write_regs(&self, vcpu_fd: &Self::Vcpu, regs: &X86_64Regs) -> Result<()>;
+
+    /// Adds hardware breakpoint
+    fn add_hw_breakpoint(&mut self, vcpu_fd: &Self::Vcpu, addr: u64) -> Result<bool> {
+        let addr = self.translate_gva(vcpu_fd, addr)?;
+
+        if self.is_hw_breakpoint(&addr) {
+            Ok(true)
+        } else {
+            let res = self.save_hw_breakpoint(&addr);
+            if res {
+                self.set_single_step(vcpu_fd, false)?;
+                Ok(true)
+            } else {
+                Ok(false)
+            }
+        }
+    }
+    /// Removes hardware breakpoint
+    fn remove_hw_breakpoint(&mut self, vcpu_fd: &Self::Vcpu, addr: u64) -> Result<bool> {
+        let addr = self.translate_gva(vcpu_fd, addr)?;
+
+        if self.is_hw_breakpoint(&addr) {
+            self.delete_hw_breakpoint(&addr);
+            self.set_single_step(vcpu_fd, false)?;
+
+            Ok(true)
+        } else {
+            Ok(false)
+        }
+    }
+}
+
+/// Trait used for defining memory debugging functionality
+pub trait GuestMemoryDebug: GuestVcpuDebug {
+    /// Returns true whether the provided address is a software breakpoint
+    fn is_sw_breakpoint(&self, addr: &u64) -> bool;
+    /// Stores the data that the sw breakpoint op code replaces
+    fn save_sw_breakpoint_data(&mut self, addr: u64, data: [u8; 1]);
+    /// Retrieves the saved data that the sw breakpoint op code replaces
+    fn delete_sw_breakpoint_data(&mut self, addr: &u64) -> Option<[u8; 1]>;
+
+    /// Overwrites the guest memory with the SW Breakpoint op code that instructs
+    /// the vCPU to stop when is executed and stores the overwritten data to be
+    /// able to restore it
+    fn add_sw_breakpoint(
```

Comment:
Same as above does the add/remove need to be `Result<bool>` or can it be `Result<()>`